RHTAPINST-137: Nexus integration with rhtap

prietyc123 opened the pull request:
Integrating nexus as container registry for rhtap

prietyc123 committed:
Nexus integration with rhtap

Roming22 commented on installer/charts/rhtap-dh/templates/extra-env.yaml:
```diff
@@ -63,3 +63,9 @@ data:
     QUAY__API_TOKEN: {{ $quaySecretData.token }}
 {{- end }}
     QUAY__URL: {{ $quaySecretData.url }}
```
You need to surround all of the QUAY vars (L62-65) with
```
{{- if $quaySecretData }}
{{- end }}

Roming22 commented on installer/charts/rhtap-dh/hooks/post-deploy.sh:
```diff
@@ -17,26 +17,56 @@ get_binaries() {
+    done
+    echo -n "."
+
+    # Check for quay-auth and nexus-auth secrets and patch if present
```
You should be able to simplify (untested):
```
    # Check for quay-auth and nexus-auth secrets and patch if present
    for SECRET_NAME in nexus-auth quay-auth; do
        SECRET=$("$KUBECTL" get secret --namespace "$NAMESPACE" --ignore-not-found)
        if [ -n "$SECRET" ]; then
            "$KUBECTL" patch serviceaccounts --namespace "$NAMESPACE" "$SA" --patch "
secrets:
    - name: $SECRET_NAME
imagePullSecrets:
    - name: $SECRET_NAME
imagePullSecrets:
" >/dev/null
        fi
        echo "OK"
    done
```

prietyc123 replied:
sure

Roming22 commented on installer/charts/rhtap-dh/templates/app-config-content.yaml:
```diff
@@ -90,11 +90,22 @@ proxy:
       changeOrigin: true
-      # Change to "false" in case of using self hosted quay instance with a self-signed certificate
+      secure: true
+    '/nexus/api':
```
You need to surround all of the quay/nexus config with
```
{{- if $****Secret}}
{{- end }
```
Same on lines 105-108

Roming22 commented on installer/charts/rhtap-dh/templates/extra-env.yaml:
```diff
@@ -63,3 +63,9 @@ data:
+{{- $nexusSecretObj := (lookup "v1" "Secret" .Release.Namespace "rhtap-nexus-integration") }}
+{{- $nexusSecretData := ($nexusSecretObj.data | default dict) }}
+{{- if $nexusSecretData }}
+    NEXUS__TOKEN: {{ $nexusSecretData.token }}
```
Are there public repositories on Nexus? If so the token might be optional. You would need to update `nexus.go` too.

prietyc123 replied:
I am also looking for this but so far couldn't find a way out.

Roming22 commented on installer/charts/values.yaml.tpl:
```diff
@@ -6,9 +6,10 @@
 {{- $gitops := required "GitOps settings" .Installer.Features.openShiftGitOps -}}
 {{- $pipelines := required "Pipelines settings" .Installer.Features.openShiftPipelines -}}
 {{- $quay := required "Quay settings" .Installer.Features.redHatQuay -}}
+{{- $quayEnabled := not (lookup "v1" "Secret" .Installer.Namespace "rhtap-nexus-integration") -}}
```
That should be `nexusEnabled`. You need to add logic to support both `quay.Enabled` and/or `nexusEnabled`.

prietyc123 replied:
IIUC we only need to worry about enable/disable quay but not nexus because we are not installing nexus operator. We only want the `rhtap-cli integrations` . Here with `quayEnabled`, I am disabling quay if `rhtap-nexus-integration` secrets are available but if not just enable quay to get it installed. Let me know if my understanding is not correct

Roming22 replied:
1. You would be missing an 'or' condition I believe. `quayEnabled` will be true if there is no nexus integration even if the user disabled quay in the config.yaml.
2. The logic you are putting in place is saying "you cannot have quay and nexus at the same time". Do we know that it is true? Is DH breaking when that is the case?

Roming22 commented on installer/charts/values.yaml.tpl:
```diff
@@ -6,9 +6,10 @@
 {{- $rhdh := required "RHDH settings" .Installer.Features.redHatDeveloperHub -}}
 {{- $ingressDomain := required "OpenShift ingress domain" .OpenShift.Ingress.Domain -}}
-{{- $minIOOperatorEnabled := or $tpa.Enabled $quay.Enabled -}}
+{{- $minIOOperatorEnabled := or $tpa.Enabled $quayEnabled -}}
```
This change looks wrong. Why do we need to install minIO if TPA is disabled and an external Nexus instance is used?

prietyc123 replied:
Here as you see it does not referring to nexus but quay. As I explained in my above comment that if `nexus` is there we are disabling quay and hence minIO should not be functional.

Roming22 replied:
Let's figure out the logic in the other thread, and then see how it impacts this.

Roming22 commented on installer/charts/values.yaml.tpl:
```diff
@@ -27,7 +28,7 @@ openshift:
     - {{ $acs.Namespace }}
 {{- end }}
-{{- if $quay.Enabled }}
+{{- if $quayEnabled }}
```
All those changes from `quay.Enabled` to `quayEnabled` look wrong to me. Maybe I missed something?

prietyc123 replied:
ptal https://github.com/redhat-appstudio/rhtap-cli/pull/247#discussion_r1794723328

Roming22 replied:
Let's figure out the logic in the other thread, and then see how it impacts this.

prietyc123 commented:
@Roming22 I addressed your review comment in [commit](https://github.com/redhat-appstudio/rhtap-cli/pull/247/commits/6f17fe92e46a1ed7f1db73309950cae55cc52964) 
Could you please take a look. Also I have responded to your concern related to enable/disable quay

Roming22 commented on installer/charts/rhtap-dh/templates/extra-env.yaml:
```diff
@@ -59,7 +59,12 @@ data:
 {{- $quaySecretObj := (lookup "v1" "Secret" .Release.Namespace "rhtap-quay-integration") }}
 {{- $quaySecretData := ($quaySecretObj.data | default dict) }}
-{{- if $quaySecretData.token }}
+{{- if $quaySecretData }}
```
The token is optional (not needed when the image repo is public).
```
{{- if $quaySecretData }}
    {{- if $quaySecretData.token }}
    QUAY__API_TOKEN: {{ $quaySecretData.token }}
    {{- end }}
    QUAY__URL: {{ $quaySecretData.url }}
{{- end }}
```

prietyc123 commented:
> The logic you are putting in place is saying "you cannot have quay and nexus at the same time". Do we know that it is true? > Is DH breaking when that is the case?

Ok I need to confirm on this. As per my understanding, I was under impression that we should not have quay and nexus at the same time

prietyc123 commented:
@Roming22 I have verified we can have quay and nexus at the same time.

Roming22 commented on installer/charts/rhtap-dh/templates/extra-env.yaml:
```diff
@@ -59,7 +59,14 @@ data:
+{{- $nexusSecretObj := (lookup "v1" "Secret" .Release.Namespace "rhtap-nexus-integration") }}
+{{- $nexusSecretData := ($nexusSecretObj.data | default dict) }}
+{{- if $nexusSecretData }}
+    NEXUS__URL: {{ $nexusSecretData.url }}
```
I'm surprised there is no token for private image repositories. OK to merge as is until we figure it out.

Roming22 commented on installer/charts/values.yaml.tpl:
```diff
@@ -6,9 +6,12 @@
 {{- $quay := required "Quay settings" .Installer.Features.redHatQuay -}}
+{{- $nexusSecret := (lookup "v1" "Secret" .Installer.Namespace "rhtap-nexus-integration") -}}
+{{- $quaySecret := (lookup "v1" "Secret" .Installer.Namespace "rhtap-quay-integration") -}}
+{{- $quayEnabled := not (or $nexusSecret $quaySecret) -}}
```
1. The variable name is confusing
2. If quay is deployed by the installer, the secret doesn't exist yet.

Roming22 replied:
Please revert

Roming22 commented on installer/charts/values.yaml.tpl:
```diff
@@ -6,9 +6,12 @@
 {{- $rhdh := required "RHDH settings" .Installer.Features.redHatDeveloperHub -}}
 {{- $ingressDomain := required "OpenShift ingress domain" .OpenShift.Ingress.Domain -}}
-{{- $minIOOperatorEnabled := or $tpa.Enabled $quay.Enabled -}}
+{{- $minIOOperatorEnabled := or $tpa.Enabled $quayEnabled -}}
```
That would install minio if there was Nexus and no TPA. Feels wrong. I don't think you need to change that condition, here or later in the file.

Roming22 commented on pkg/subcmd/integration_nexus.go:
```diff
@@ -0,0 +1,98 @@
+var _ Interface = &IntegrationNexus{}
+
+const nexusIntegrationLongDesc = `
+Manages the nexus integration with RHTAP, by storing the required
```
Nit pick: Use 'Nexus' when referring to the product name instead of 'nexus'.

Roming22 commented on pkg/subcmd/integration_nexus.go:
```diff
@@ -0,0 +1,98 @@
+for RHDH.
+
+
+The given dockerconfig must include the repository path.
```
Remove that line. I believe it was for a quay issue.

Roming22 commented on installer/config.yaml:
```diff
@@ -19,7 +19,7 @@ rhtapCLI:
         catalogURL: https://github.com/redhat-appstudio/tssc-sample-templates/blob/release/all.yaml
     redHatAdvancedClusterSecurity:
-      enabled: *rhdhEnabled
+      enabled: false
```
Please revert.

Roming22 reviewed (approved):
/lgtm

Roming22 commented:
/retest

prietyc123 commented:
/retest

prietyc123 commented:
/retest

prietyc123 commented:
/retest

xinredhat commented:
/lgtm

prietyc123 commented:
/retest

xinredhat commented:
/lgtm

Roming22 commented:
/retest

prietyc123 commented:
/retest

xinredhat commented:
/lgtm
